src/keys.rs

```diff
@@ -1,70 +1,193 @@
+        CtOption::new(sk, !ask.ct_is_zero())
+    }
+}
 
+/// A spending authorizing key, used to create spend authorization signatures.
+///
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][§4.2.3].
+///
+/// [§4.2.3]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
 #[derive(Debug)]
-pub(crate) struct SpendAuthorizingKey;
+pub(crate) struct SpendAuthorizingKey(redpallas::SigningKey<SpendAuth>);
+
+impl SpendAuthorizingKey {
+    /// Derives ask from sk. Internal use only, does not enforce all constraints.
+    fn derive_inner(sk: &SpendingKey) -> pallas::Scalar {
+        to_scalar(prf_expand(&sk.0, &[0x06]))
+    }
+}
 
 impl From<&SpendingKey> for SpendAuthorizingKey {
-    fn from(_: &SpendingKey) -> Self {
-        todo!()
+    fn from(sk: &SpendingKey) -> Self {
+        let ask = Self::derive_inner(sk);
+        // SpendingKey cannot be constructed such that this assertion would fail.
+        assert!(!bool::from(ask.ct_is_zero()));
+        // TODO: Add TryFrom<S::Scalar> for SpendAuthorizingKey.
+        let ret = SpendAuthorizingKey(ask.to_bytes().try_into().unwrap());
+        // If the last bit of repr_P(ak) is 1, negate ask.
+        if (<[u8; 32]>::from(AuthorizingKey::from(&ret).0)[31] >> 7) == 1 {
+            SpendAuthorizingKey((-ask).to_bytes().try_into().unwrap())
+        } else {
+            ret
+        }
     }
 }
 
 /// TODO: This is its protocol spec name for Sapling, but I'd prefer a different name.
 #[derive(Debug)]
-pub(crate) struct AuthorizingKey;
+pub(crate) struct AuthorizingKey(redpallas::VerificationKey<SpendAuth>);
 
 impl From<&SpendAuthorizingKey> for AuthorizingKey {
-    fn from(_: &SpendAuthorizingKey) -> Self {
-        todo!()
+    fn from(ask: &SpendAuthorizingKey) -> Self {
+        AuthorizingKey((&ask.0).into())
     }
 }
 
+/// A key used to derive [`Nullifier`]s from [`Note`]s.
+///
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][§4.2.3].
+///
+/// [`Nullifier`]: crate::note::Nullifier;
+/// [`Note`]: crate::note::Note;
+/// [§4.2.3]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
 #[derive(Debug)]
-pub(crate) struct NullifierDerivingKey;
+pub(crate) struct NullifierDerivingKey(pallas::Base);
 
 impl From<&SpendingKey> for NullifierDerivingKey {
-    fn from(_: &SpendingKey) -> Self {
-        todo!()
+    fn from(sk: &SpendingKey) -> Self {
+        NullifierDerivingKey(to_base(prf_expand(&sk.0, &[0x07])))
     }
 }
 
 /// A key that provides the capability to view incoming and outgoing transactions.
 ///
 /// This key is useful anywhere you need to maintain accurate balance, but do not want the
 /// ability to spend funds (such as a view-only wallet).
-///
-/// TODO: Should we just define the FVK to include extended stuff like the diversifier key?
 #[derive(Debug)]
 pub struct FullViewingKey {
     ak: AuthorizingKey,
     nk: NullifierDerivingKey,
-    rivk: (),
+    rivk: pallas::Scalar,
```

Comment:
Perhaps a wrapper type, `IvkCommitRandomness`, that has
```rust
struct IvkCommitRandomness(pallas::Scalar);

impl From<SpendingKey> for IvkCommitRandomness {
    pub from(sk: &SpendingKey) - > Self {
        Self(pallas::Scalar::from_bytes_wide(prf_expand(&sk.0, &[0x08])))
    }
}
```
?
or using the `to_scalar` function

Reply:
Isn't it already using `to_scalar`?

Reply:
Confirmed it is. But yeah I can add that as another internal type.